cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/ConsumerWorkingSetOperations.java

```diff
@@ -0,0 +1,181 @@
+    PartitionConsumerResult result = partitionedFileSet.consumePartitions(workingSet.getPartitionConsumerState(),
+                                                                          numToPopulate, predicate);
+    List<PartitionDetail> partitions = result.getPartitions();
+    for (PartitionDetail partition : partitions) {
+      workingSet.getPartitions().add(new ConsumablePartition(partition.getPartitionKey()));
+    }
+    workingSet.setPartitionConsumerState(result.getPartitionConsumerState());
+  }
+
+  /**
+   * This method must be called on any partitions returned by the {@code #consumePartitions} method.
+   *
+   * @param partitionKeys list of partition keys to mark as either succeeded or failed processing
+   * @param succeeded whether or not processing of the specified partitions was successful
+   */
+  public void onFinish(List<PartitionKey> partitionKeys, boolean succeeded) {
+    // TODO: user must return them in the same order that it was given to them in
+    // reason: in order to optimize commit/abort methods, which is also currently a TODO
+    if (succeeded) {
+      commit(partitionKeys);
+    } else {
+      abort(partitionKeys);
+    }
+  }
+
+  /**
+   * Removes the given partition keys from the working set, as they have been successfully processed.
+   */
+  private void commit(List<PartitionKey> partitionKeys) {
+    long expiryTime = getExpiryBorder();
+
+    for (PartitionKey key : partitionKeys) {
+      ConsumablePartition consumablePartition = workingSet.lookup(key);
+      assertInProgress(consumablePartition);
+      assertNotExpired(consumablePartition, expiryTime);
+      workingSet.remove(key);
+    }
+  }
+
+  /**
+   * Resets the process state of the given partition keys, as they were not successfully processed.
+   */
+  private void abort(List<PartitionKey> partitionKeys) {
+    long expiryTime = getExpiryBorder();
+
+    for (PartitionKey key : partitionKeys) {
+      ConsumablePartition consumablePartition = workingSet.lookup(key);
+      assertInProgress(consumablePartition);
+      assertNotExpired(consumablePartition, expiryTime);
+      consumablePartition.resetState();
+    }
+  }
+
+  // ensure that user doesn't try to commit/abort a partition that isn't in progress
+  private void assertInProgress(ConsumablePartition consumablePartition) {
+    if (!consumablePartition.isInProgress()) {
+      throw new IllegalArgumentException(String.format("Partition not in progress: %s",
+                                                       consumablePartition.getPartitionKey()));
+    }
+  }
+
+  // ensure that user doesn't try to commit/abort a partition that is expired
+  private void assertNotExpired(ConsumablePartition consumablePartition, long expiryBorder) {
+    if (consumablePartition.getTimestamp() < expiryBorder) {
+      throw new IllegalArgumentException(String.format("Attempting to operate on an expired partition: %s",
```

Comment:
Is this an illegal argument? This happens if processing takes too long. So it is expected some times. Shouldn't it then throw a checked exception?